File: Cpp/Kex_2_Battle.cpp
#include "Kex_2_Battle.h"
#include <iostream>
#include <random>

using std::cout;

bool Kex_2_Battle::CheckHitOrMiss(int hit)
{
	auto engine = Kex_2::GetEngine<>();
	std::uniform_int_distribution<> generator(0, 100);
	int value = generator(engine);
	return value <= hit - 1;
}

int Kex_2_Battle::GetAttackPower()
{
	auto engine = Kex_2::GetEngine<>();
	std::uniform_int_distribution<> generator(0, 40);
	
	return generator(engine) + 60;
}

int Kex_2_Battle::GetSkillPower()
{
	auto engine = Kex_2::GetEngine<>();
	std::uniform_int_distribution<> generator(0, 100);

	return generator(engine) + 30;
}

int Kex_2_Battle::GetMagicPower()
{
	auto engine = Kex_2::GetEngine<>();
	std::uniform_int_distribution<> generator(0, 180);

	return generator(engine) + 20;
}

bool Kex_2_Battle::AttackEnemy(PlayerEx& player, Enemy& enemy)
{
	cout << enemy.Name << "Lv." << enemy.Lv + 1 << "が現れた！\n";

	while (enemy.Hp > 0)
	{
		cout << enemy.Name << " 残りHP：" << enemy.Hp << "\n";
		cout << "武器を選択してください（１．攻撃　２．特技　３．魔法）＞";
		int choice = 0;
		std::cin >> choice;
		int damage = 0;
		bool is_hit = false;
		int (*get_powers[])() = { GetAttackPower, GetSkillPower, GetMagicPower };
		int get_hits[] = { ATTACK_HIT, SKILL_HIT, MAGIC_HIT };
		Attack attacks[3] = {};
		for (int i = 0; i < 3; ++i)
		{
			attacks[i].Damage = get_powers[i]();
			attacks[i].Hit = get_hits[i];
		}

		switch (choice)
		{
		case 1:
			damage = attacks[0].Damage;
			is_hit = CheckHitOrMiss(attacks[0].Hit - enemy.Flee);
			break;
		case 2:
			damage = attacks[1].Damage;
			is_hit = CheckHitOrMiss(attacks[1].Hit - enemy.Flee);
			break;
		case 3:
			damage = attacks[2].Damage;
			is_hit = CheckHitOrMiss(attacks[2].Hit - enemy.Flee);
			break;
		}

		if (is_hit)
		{
			damage -= enemy.Defense;
			if (damage <= 0)
				damage = 0;
			cout << damage << "のダメージ！\n";
			enemy.Hp -= damage;
		}
		else
		{
			cout << "攻撃を外した！\n";
		}

		cout << enemy.Name << "の攻撃！\n";
		is_hit = CheckHitOrMiss(enemy.Hit);
		if (is_hit)
		{
			int injury = enemy.Attack - player.Defense;
			if (injury <= 0)
				injury = 0;
			cout << injury << "のダメージ！\n";
			player.Hp -= injury;
			if (player.Hp <= 0)
			{
				player.Hp = 0;
				cout << "あなたは" << enemy.Name << "に負けました！ゲームオーバー！\n";
				return false;
			}
		}
		else
		{
			cout << "攻撃を外した！\n";
		}

		cout << "プレイヤー残りHP：" << player.Hp << "\n";

		if (enemy.Hp <= 0)
			enemy.Hp = 0;
		if (enemy.Hp == 0)
			cout << enemy.Name << "Lv." << enemy.Lv + 1 << "を倒した！\n";
	}

	return true;
}

File: Go/Kex_2_battle.go
package main

import (
	"fmt"
	"math/rand"
	"strconv"
)

func AttackEnemy(player *Player, enemy *Enemy) bool {
	fmt.Printf("%sLv.%dが現れた！\n", enemy.Name, enemy.Lv + 1)

	for enemy.Hp > 0 {
		fmt.Printf("%s 残りHP：%d\n", enemy.Name, enemy.Hp)
		fmt.Print("武器を選択してください（１．攻撃　２．特技　３．魔法）＞")
		var reader string
		_, _ = fmt.Scanln(&reader)
		choice, _ := strconv.Atoi(reader)
		getAttackPower := func () int {
			return rand.Intn(41) + 60
		}
		getSkillPower := func () int {
			return rand.Intn(101) + 30
		}
		getMagicPower := func () int {
			return rand.Intn(181) + 20
		}
		getPowers := []func () int { getAttackPower, getSkillPower, getMagicPower }
		var attacks []Attack
		getHits := []int { AttackHit, SkillHit, MagicHit }

		for i := 0; i < 3; i++ {
			attacks = append(attacks, Attack { getPowers[i](), getHits[i] })
		}

		damage := 0
		isHit := false

		switch choice {
		case 1:
			damage = attacks[0].Damage
			isHit = CheckHitOrMiss(attacks[0].Hit - enemy.Flee)
		case 2:
			damage = attacks[1].Damage
			isHit = CheckHitOrMiss(attacks[1].Hit - enemy.Flee)
		case 3:
			damage = attacks[2].Damage
			isHit = CheckHitOrMiss(attacks[2].Hit - enemy.Flee)
		}

		if isHit {
			damage -= enemy.Defense
			if damage <= 0 {
				damage = 0
			}
			fmt.Printf("%dのダメージ！\n", damage)
			enemy.Hp -= damage
		} else {
			fmt.Println("攻撃を外した！")
		}

		fmt.Printf("%sの攻撃！\n", enemy.Name)
		isHit = CheckHitOrMiss(enemy.Hit)
		if isHit {
			injury := enemy.Attack - player.Defense
			if injury <= 0 {
				injury = 0
			}
			fmt.Printf("%dのダメージ！\n", injury)
			player.Hp -= injury
			if player.Hp <= 0 {
				player.Hp = 0
				fmt.Printf("あなたは%sに負けました！ゲームオーバー！\n", enemy.Name)
				return false
			}
		} else {
			fmt.Println("攻撃を外した！")
		}

		fmt.Printf("プレイヤー残りHP：%d\n", player.Hp)
		if enemy.Hp <= 0 {
			enemy.Hp = 0
		}
		if enemy.Hp == 0 {
			fmt.Printf("%sLv.%dを倒した！\n", enemy.Name, enemy.Lv + 1)
		}
	}
	return true
}